activesupport/lib/active_support/duration.rb

```diff
@@ -125,5 +183,52 @@ def sum(sign, time = ::Time.current) #:nodoc:
       def method_missing(method, *args, &block) #:nodoc:
         value.send(method, *args, &block)
       end
+
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601DurationParser
+      attr_reader :parts
+
+      class ParsingError < ::StandardError; end
+
+      def initialize(iso8601duration)
+        match = iso8601duration.match(/^
+              (?<sign>\+|-)?
+              P(?:
+                (?:
+                  (?:(?<years>-?\d+(?:[,.]\d+)?)Y)?
+                  (?:(?<months>-?\d+(?:[.,]\d+)?)M)?
+                  (?:(?<days>-?\d+(?:[.,]\d+)?)D)?
+                  (?<time>T
+                    (?:(?<hours>-?\d+(?:[.,]\d+)?)H)?
+                    (?:(?<minutes>-?\d+(?:[.,]\d+)?)M)?
+                    (?:(?<seconds>-?\d+(?:[.,]\d+)?)S)?
+                  )?
+                ) |
+                (?<weeks>-?\d+(?:[.,]\d+)?W)
+              ) # Duration
+            $/x) || raise(ParsingError.new("Invalid ISO 8601 duration: #{iso8601duration}"))
+        sign = match[:sign] == '-' ? -1 : 1
+        @parts = match.names.zip(match.captures).reject{|_k,v| v.nil? }.map do |k, v|
```

Comment:
I'd extract this to a private method. Also we prefer spaces in blocks `reject { |_, v| v.nil? }`.